h2/utilities.py

```diff
@@ -242,3 +249,41 @@ def _reject_pseudo_header_fields(headers):
             seen_regular_header = True
 
         yield header
+
+
+def _check_host_authority_header(headers, connection_state):
+    """
+    Raises a ProtocolError if a header block arrives that does not contain
+    :authority or a Host header, or if a header block contains both fields,
+    but their values do not match.
+    """
+    # We only expect to see :authority and Host headers on request header
+    # blocks that aren't trailers, so skip this validation if we're on the
+    # server side or looking at trailer blocks.
+    if connection_state.is_client or connection_state.is_trailer:
+        for header in headers:
+            yield header
+        return
+
+    authority_header_val = None
+    host_header_val = None
+
+    for header in headers:
+        if header[0] == b':authority':
+            authority_header_val = header[1]
+        elif header[0] == b'host':
+            host_header_val = header[1]
+
+        yield header
+
+    if (authority_header_val is None) and (host_header_val is None):
+        raise ProtocolError(
+            "Did not receive an :authority or Host header."
+        )
+
+    if (authority_header_val is not None) and (host_header_val is not None):
+        if authority_header_val != host_header_val:
+            raise ProtocolError(
+                "Received mismatched :authority and Host headers: %r / %r" %
+                (authority_header_val, host_header_val)
+            )
```

Comment:
This doesn't have to be dealt with in this PR, but we'll also want to reject this kind of header block when our users send it. That means we'll probably want to factor the core of this logic (checking the header block) out to a function we can call in a separate situation to this one. You can either do that now, or we can do it in a follow-up PR: your call.